### PR TITLE
Handle client error

### PR DESCRIPTION
**Pull request by delvedor**
Do we need to log this?
Feedbacks?

Rel: #102

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -358,6 +360,16 @@ function build (options) {
     return this._contentTypeParser.hasParser(contentType)
   }
 
+  function handleClientError (e, socket) {
+    socket.end(
+      JSON.stringify({
```
Comment (allevo) [inline fastify.js]: This can be expensive: could it be built on the fastify start up keeping the string in memory?

Comment (delvedor) [inline fastify.js]: Yes we can, but I don't think we will have so many client errors! :D

File: fastify.js [hunk 2]
```diff
@@ -358,6 +360,16 @@ function build (options) {
...
+        error: http.STATUS_CODES['400'],
+        message: 'Client Error',
+        statusCode: 400
+      })
```
Comment (allevo) [inline fastify.js]: According to documentation, the socket should be ended with a valid http response.
Here, the socket will be closed with a json output.
Maybe a simple "HTTP/1.1 400 Bad Request\r\n\r\n" can be used

Comment (delvedor) [inline fastify.js]: I thought about that, but I like to be consistent with the [standard error](https://github.com/fastify/fastify/blob/master/docs/Reply.md#errors) we use across Fastify:
```js
{
  error: String        // the http error message
  message: String      // the user error message
  statusCode: Number   // the http status code
}
```

Comment (allevo) [inline fastify.js]: Yes, using the HTTP protocol, the response body should be conformed with the standard.
But in that callback the second parameter is a tcp socket not a http response.
So ending the socket with a body like this, is better

    const body = JSON.stringify({
      error: http.STATUS_CODES['400'],
      message: 'Client Error',
      statusCode: 400
    })
    socket.end(`HTTP/1.1 400 Bad Request\r\nContent-Length: ${body.length}\r\n\r\n${body}`)

Comment (mcollina) [inline fastify.js]: :+1: on @allevo suggestion.

File: fastify.js [hunk 3]
```diff
@@ -358,6 +362,15 @@ function build (options) {
     return this._contentTypeParser.hasParser(contentType)
   }
 
+  function handleClientError (e, socket) {
+    const body = JSON.stringify({
+      error: http.STATUS_CODES['400'],
+      message: 'Client Error',
+      statusCode: 400
+    })
+    socket.end(`HTTP/1.1 400 Bad Request\r\nContent-Length: ${body.length}\r\n\r\n${body}`)
```
Comment (stefanduberg) [inline fastify.js]: Should perhaps set `content-type: application/json` header here as well as the response body is JSON.

File: fastify.js [hunk 4]
```diff
@@ -358,6 +362,15 @@ function build (options) {
...
   }
 
+  function handleClientError (e, socket) {
+    const body = JSON.stringify({
+      error: http.STATUS_CODES['400'],
+      message: 'Client Error',
+      statusCode: 400
+    })
+    socket.end(`HTTP/1.1 400 Bad Request\r\nContent-Length: ${body.length}\r\nContent-Type: 'application/json'\r\n\r\n${body}`)
+  }
```
Comment (mcollina) [inline fastify.js]: You are right, we should log this.

**Pull request merged**